pytools/persistent_dict.py

```diff
@@ -301,6 +315,10 @@ def update_for_int(key_hash, key):
                 return
             except OverflowError:
                 sz *= 2
+            except AttributeError:
+                # Numpy scalars don't have to_bytes()
```

Comment:
I think a better approach would be to have all numpy scalars go to the same method, in which you convert to shape-`()` array and then use `to_bytes` on that.

Reply:
Should we add similar functionality for ndarrays as well? maybe something like (tobytes, shape, dtype)?

Reply:
No, definitely not!
- They're mutable.
- With that implementation, it would violate the fundamental property a hash satisfies, i.e. `a == b` imples `hash(a) == hash(b)`. That's because the array's byte pattern depends on storage order (e.g. F/C ordering), for arrays that Numpy would otherwise consider equal.

Reply:
It seems that `tobytes()` for complex256 (only that dtype) isn't stable:

```python
import numpy as np

for i in range(10):
    k = np.complex256(1.1+2.2j)
    print(k.tobytes())
```

```
$ python t.py
b'\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\xff?\x00\x00\x00\x00\x00\x00\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\x00@\x00\x00\x00\x00\x00\x00'
b'\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\xff?\x00\x00\x00\x00\x00\x00\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\x00@\x00\x00\x00\x00\x00\x00'
b'\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\xff?\x00\x00\x00\x00\x00\x00\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\x00@\x00\x00\x00\x00\x00\x00'
b'\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\xff?\x00\x00\x00\x00\x00\x00\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\x00@\x00\x00\x00\x00\x00\x00'
b'\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\xff?\x00\x00\x00\x00\x00\x00\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\x00@\x00\x00\x00\x00\x00\x00'
b'\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\xff?\x00\x00\x00\x00\x00\x00\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\x00@\x00\x00\x00\x00\x00\x00'
b'\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\xff?\x00\x00\x00\x00\x00\x00\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\x00@\x00\x00\x00\x00\x00\x00'
b'\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\xff?\x00\x00\x00\x00\x00\x00\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\x00@\x00\x00\x00\x00\x00\x00'
b'\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\xff?\x86\xc5\xeel\x16\xfe\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\x00@`w\x00\x00\x00\x00'
b'\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\xff?\x86\xc5\xeel\x16\xfe\x00\xd0\xcc\xcc\xcc\xcc\xcc\x8c\x00@`w\x00\x00\x00\x00'
```

Reply:
That makes a little bit of sense to me, as (at least on x86), `float128` (where `complex256` is just two of those) is just the x87 (not a typo) 80-bit FP type. So there are six bytes that aren't accounted for and possibly arbitrary.

Maybe specially handle those? (Cast to `float` (yes, Python `float`, both parts for complex), call `repr` and hash that)?

Reply:
> That makes a little bit of sense to me, as (at least on x86), `float128` (where `complex256` is just two of those) is just the x87 (not a typo) 80-bit FP type. So there are six bytes that aren't accounted for and possibly arbitrary.

Oh 😞  In other words, this might also affect `float128` itself?

> Maybe specially handle those? (Cast to `float` (yes, Python `float`, both parts for complex), call `repr` and hash that)?

If you prefer that idea to https://github.com/inducer/pytools/pull/184/commits/f1b766c791299b08bf4f7090b9b47b4dbfe2678a, I can implement that.

Reply:
I implemented your suggestion in https://github.com/inducer/pytools/pull/184/commits/aeb824ad94a4032d64acfecd3affe7edcb7a88b7 (using Python's `complex()` for converting `np.complex256` because that seemed a bit simper).